buffer/src/main/java/io/netty/buffer/api/internal/ArcDrop.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Copyright 2021 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.buffer.api.internal;
+
+import io.netty.buffer.api.Drop;
+
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.VarHandle;
+
+public final class ArcDrop<T> implements Drop<T> {
+    private static final VarHandle COUNT;
+    static {
+        try {
+            COUNT = MethodHandles.lookup().findVarHandle(ArcDrop.class, "count", int.class);
+        } catch (Exception e) {
+            throw new ExceptionInInitializerError(e);
+        }
+    }
+
+    private final Drop<T> delegate;
+    @SuppressWarnings("FieldMayBeFinal")
+    private volatile int count;
+
+    public ArcDrop(Drop<T> delegate) {
+        this.delegate = delegate;
+        count = 1;
+    }
+
+    public static <X> Drop<X> wrap(Drop<X> drop) {
+        if (drop.getClass() == ArcDrop.class) {
+            return drop;
+        }
+        return new ArcDrop<>(drop);
+    }
+
+    public static <X> Drop<X> acquire(Drop<X> drop) {
+        if (drop.getClass() == ArcDrop.class) {
+            ((ArcDrop<X>) drop).increment();
+            return drop;
+        }
+        return new ArcDrop<>(drop);
+    }
+
+    public ArcDrop<T> increment() {
+        int c;
+        do {
+            c = count;
+            checkValidState(c);
+        } while (!COUNT.compareAndSet(this, c, c + 1));
```

Comment:
Not sure where this is used (GitHub is wheezing on this PR) but If this is the primary drop implementation I think you should consider being more loose with these increments.  There was a performance PR I made a while back that sped up the ref counting by making it non deterministic which thread cleans up in the event of double free.   

The change I made ended up getting reverted (and then eventually rewritten by @njhill ) because some project was depending on precise ref counting semantics and exceptions as flow control.    Since this new API has no semantics associated with it yet, consider setting a new precedant.

In particular, this can use getAndIncrement, which is wait free on x86.

Reply:
This is used when a buffer is `split`. The increment and the state check needs to be atomic. We can't be relaxed here in case we're racing with a `close` call.